src/NHibernate/Type/AbstractDateTimeType.cs

```diff
@@ -0,0 +1,160 @@
+
+		/// <inheritdoc />
+		public override System.Type ReturnedClass => typeof(DateTime);
+
+		/// <summary>
+		/// Retrieve the current system time.
+		/// </summary>
+		/// <value><see cref="DateTime.UtcNow" /> if <see cref="Kind" /> is <see cref="DateTimeKind.Utc" />,
+		/// <see cref="DateTime.Now" /> otherwise.</value>
+		protected virtual DateTime Now => Kind == DateTimeKind.Utc ? DateTime.UtcNow : DateTime.Now;
+
+		/// <summary>
+		/// Default constructor.
+		/// </summary>
+		protected AbstractDateTimeType() : base(SqlTypeFactory.DateTime)
+		{
+		}
+
+		/// <summary>
+		/// Constructor for overriding the default <see cref="SqlType"/>.
+		/// </summary>
+		/// <param name="sqlTypeDateTime">The <see cref="SqlType"/> to use.</param>
+		protected AbstractDateTimeType(SqlType sqlTypeDateTime) : base(sqlTypeDateTime)
+		{
+		}
+
+		/// <summary>
+		/// Adjust the date time value for this type from an arbitrary date time value.
+		/// </summary>
+		/// <param name="dateValue">The <see cref="System.DateTime" /> to adjust.</param>
+		/// <returns>A <see cref="System.DateTime" />.</returns>
+		protected virtual DateTime AdjustDateTime(DateTime dateValue) =>
+			Kind == DateTimeKind.Unspecified ? dateValue : DateTime.SpecifyKind(dateValue, Kind);
+
+		/// <inheritdoc />
+		public override object Get(DbDataReader rs, int index, ISessionImplementor session) =>
+			GetDateTime(rs, index, session);
+
+		/// <inheritdoc />
+		public override object Get(DbDataReader rs, string name, ISessionImplementor session) =>
+			Get(rs, rs.GetOrdinal(name), session);
+
+		/// <summary>
+		/// Get the <see cref="DateTime" /> in the <see cref="DbDataReader"/> for the Property.
+		/// </summary>
+		/// <param name="rs">The <see cref="DbDataReader"/> that contains the value.</param>
+		/// <param name="index">The index of the field to get the value from.</param>
+		/// <param name="session">The session for which the operation is done.</param>
+		/// <returns>An object with the value from the database.</returns>
+		protected virtual DateTime GetDateTime(DbDataReader rs, int index, ISessionImplementor session)
+		{
+			try
+			{
+				return AdjustDateTime(Convert.ToDateTime(rs[index]));
+			}
+			catch (Exception ex)
+			{
+				throw new FormatException($"Input string '{rs[index]}' was not in the correct format.", ex);
+			}
+		}
+
+		/// <inheritdoc />
+		public override void Set(DbCommand st, object value, int index, ISessionImplementor session) =>
+			st.Parameters[index].Value = GetDateTimeToSet(value, session);
+
+		/// <summary>
+		/// Get the <see cref="DateTime" /> from the value to set in a parameter.
+		/// </summary>
+		/// <param name="value">The original value.</param>
+		/// <param name="session">The session for which the operation is done.</param>
+		/// <returns>The <see cref="DateTime" /> to set as parameter value.</returns>
+		protected virtual DateTime GetDateTimeToSet(object value, ISessionImplementor session) =>
```

Comment:
Overridden by `TimestampType` for its old logic replacing non datetime values by `Now`. Intended to replace `null` it seems, it was never doing that since `Set` is never called on `null` values. But it may replace invalid values set through a property typed `object` .